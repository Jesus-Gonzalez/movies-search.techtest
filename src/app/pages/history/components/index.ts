export { default as HistoryListContainer } from './HistoryListContainer';
