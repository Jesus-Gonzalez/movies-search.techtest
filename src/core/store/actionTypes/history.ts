export const History = {
  Add: 'HISTORY_ADD',
  Remove: 'HISTORY_REMOVE',
  Clear: 'HISTORY_CLEAR'
};
