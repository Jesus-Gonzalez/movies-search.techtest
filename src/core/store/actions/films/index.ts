export * from './fetch';
export * from './filter';
export * from './get';
export * from './setLoading';
export * from './submit';
export * from './toggleSuggestions';
