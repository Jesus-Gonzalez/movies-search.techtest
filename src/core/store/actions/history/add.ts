export const addToHistory = (film: IFilm) => ({
  type: 'HISTORY_ADD',
  film,
});
