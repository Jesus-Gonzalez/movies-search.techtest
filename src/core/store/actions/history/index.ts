export * from './add';
export * from './clear';
export * from './remove';
