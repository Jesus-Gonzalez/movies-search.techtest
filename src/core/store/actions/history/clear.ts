import * as ActionTypes from 'core/store/actionTypes';

export const clear = () => ({
  type: ActionTypes.History.Clear,
});
