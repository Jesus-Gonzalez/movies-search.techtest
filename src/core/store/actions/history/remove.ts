import * as ActionTypes from 'core/store/actionTypes';

export const remove = (film: IFilmObject) => ({
  type: ActionTypes.History.Remove,
  film,
});
