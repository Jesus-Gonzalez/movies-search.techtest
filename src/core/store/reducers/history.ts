import * as ActionTypes from 'core/store/actionTypes';

export interface IHistoryReducerState {
  items: IFilmObject[];
}

const history = JSON.parse(localStorage.getItem('history.items') || '[]');
const initialState: IHistoryReducerState = {
  items: history,
};

export default (state = initialState, action: any) => {
  switch (action.type) {
    case ActionTypes.History.Add:
      return Object.assign(
        {},
        state,
        { items: state.items.concat(action.film) }
      );

    case ActionTypes.History.Remove:
      const index = state.items.findIndex(film => film.url === action.film.url);
      const nextItems = Object.assign([], state.items);
      nextItems.splice(index, 1);

      return Object.assign(
        {},
        state,
        { items: nextItems }
      );

    case ActionTypes.History.Clear:
      return Object.assign(
        {},
        state,
        { items: [] }
      );

    default:
      return state;
  }
}
