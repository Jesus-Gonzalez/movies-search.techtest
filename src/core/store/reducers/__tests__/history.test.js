// every action type has the same pattern

import { addToHistory, remove, clear } from 'core/store/actions/history'
import historyReducer from '../history';

it('processes actions with a processing label', () => {
  const a = { id: 123, name: 'hello world', url: 'a' };
  const b = { id: 321, name: 'hello world', url: 'b' };
  let state = { items: [] };

  state = historyReducer(state, addToHistory(a));
  expect(state).toMatchSnapshot();

  state = historyReducer(state, addToHistory(b));
  expect(state).toMatchSnapshot();

  state = historyReducer(state, remove(a));
  expect(state).toMatchSnapshot();

  state = historyReducer(state, clear());
  expect(state).toMatchSnapshot();
});
