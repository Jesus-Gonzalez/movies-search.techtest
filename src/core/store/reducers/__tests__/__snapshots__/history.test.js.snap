// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processes actions with a processing label 1`] = `
Object {
  "items": Array [
    Object {
      "id": 123,
      "name": "hello world",
      "url": "a",
    },
  ],
}
`;

exports[`processes actions with a processing label 2`] = `
Object {
  "items": Array [
    Object {
      "id": 123,
      "name": "hello world",
      "url": "a",
    },
    Object {
      "id": 321,
      "name": "hello world",
      "url": "b",
    },
  ],
}
`;

exports[`processes actions with a processing label 3`] = `
Object {
  "items": Array [
    Object {
      "id": 321,
      "name": "hello world",
      "url": "b",
    },
  ],
}
`;

exports[`processes actions with a processing label 4`] = `
Object {
  "items": Array [],
}
`;
